diff --git a/drivers/spi/spidev.c b/drivers/spi/spidev.c
index e3c19f3..158cf0c 100644
--- a/drivers/spi/spidev.c
+++ b/drivers/spi/spidev.c
@@ -695,6 +695,8 @@ static struct class *spidev_class;
 static const struct of_device_id spidev_dt_ids[] = {
 	{ .compatible = "rohm,dh2228fv" },
 	{ .compatible = "lineartechnology,ltc2488" },
+	{ .compatible = "allwinner,sun4i-a10-spi" },
+	{ .compatible = "allwinner,sun6i-a31-spi" },
 	{},
 };
 MODULE_DEVICE_TABLE(of, spidev_dt_ids);
